pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/source/TopicSchema.java

```diff
@@ -104,20 +101,41 @@ private SchemaType getSchemaTypeOrDefault(String topic, Class<?> clazz) {
     }
 
     private static SchemaType getDefaultSchemaType(Class<?> clazz) {
-        if (byte[].class.equals(clazz)
-            || ByteBuf.class.equals(clazz)
-            || ByteBuffer.class.equals(clazz)) {
-            return SchemaType.NONE;
+        if (byte[].class.equals(clazz)) {
+            return BytesSchema.of().getSchemaInfo().getType();
+        } else if (ByteBuf.class.equals(clazz)) {
+            return ByteBufSchema.of().getSchemaInfo().getType();
+        } else if (ByteBuffer.class.equals(clazz)) {
+            return ByteBufferSchema.of().getSchemaInfo().getType();
+        } else if (Byte.class.equals(clazz)) {
+            return ByteSchema.of().getSchemaInfo().getType();
         } else if (GenericRecord.class.isAssignableFrom(clazz)) {
             // the function is taking generic record, so we do auto schema detection
             return SchemaType.AUTO_CONSUME;
         } else if (String.class.equals(clazz)) {
-            // If type is String, then we use schema type string, otherwise we fallback on default schema
             return SchemaType.STRING;
         } else if (isProtobufClass(clazz)) {
             return SchemaType.PROTOBUF;
         } else if (KeyValue.class.equals(clazz)) {
             return SchemaType.KEY_VALUE;
+        } else if (Date.class.equals(clazz)) {
```

Comment:
I am not sure about the branches below. I think it is a breaking change. because these types were treated as DEFAULT_SCHEMA_TYPE before, then you are changing them to a different schema. this will fail functions when runtime is upgraded to include this change.

unless I miss something around here.

Reply:
Will do the other changes, but wanted to discuss this.
IMO the current behaviour of functions wrt these is buggy. When I output Date for instance in my function, I expect it to go out as Date and not as bytes as it does right now. This means that all downstream processes(whether functions/sql, etc) will have no other way to interpret this data other than bytes. This basically renders functions useless for a lot of use cases.

Reply:
1) the default is JSON.
2) I agree it is buggy. But there were functions running before. You can’t break them. We have been adding quite a few changes that break users. I would prefer we have a solution for BC. Otherwise it looks really bad. Especially this change is about changing schema. I would prefer making them compatible.
3) if people want to use Date, he can specify schema type or precreate a topic with Date schema. Correct?